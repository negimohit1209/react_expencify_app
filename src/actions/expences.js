import uuid from "uuid";
import database from '../firebase/firebase';
// ADD_EXPENSE
export const addExpence = (expence) => ({
  type: "ADD_EXPENCE",
  expence
});

export const startAddExpence = (expenceData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenceData;
    const expence = {
      description,note,amount,createdAt
    }
    database.ref('expences').push(expence).then((ref) => {
      dispatch(addExpence({
        id: ref.key,
        ...expence
      })); 
    });
  };
};
// REMOVE_EXPENSE
export const removeExpence = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});
// EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});

//SET_EXPENCES
export const setExpences = (expences) => ({
  type: "SET_EXPENCES",
  expences
});

export const startSetExpences = () => {
  return (dispatch) => {
    return database.ref('expences').once('value').then((snapshot) => {
      const expences = [];
      snapshot.forEach((childSnapshot) => {
        expences.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpences(expences));
    });
    
  };
};
