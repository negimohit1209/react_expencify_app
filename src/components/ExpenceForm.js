import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import "react-dates/lib/css/_datepicker.css";

// const date = new Date();
const date = moment();
console.log(date.format("MMM Do, YYYY"));
export default class ExpenceForm extends Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expence? props.expence.description : "",
            note: props.expence? props.expence.note : "",
            amount: props.expence? (props.expence.amount/100).toString() : "",
            createdAt: props.expence? moment(props.expence.createdAt) : moment(),
            calenderFocused: false,
            error: ""
          };
    }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    if(!this.state.description || !this.state.amount){
        this.setState(() => ({error: "Please fill the description and amount."}))
    }else{
        this.setState(() => ({error: ""}));
        this.props.onSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100,
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
        })
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="test"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calenderFocused} // PropTypes.bool
            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
            id="your_unique_id" // PropTypes.string.isRequired,
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add expence</button>
        </form>
      </div>
    );
  }
}
