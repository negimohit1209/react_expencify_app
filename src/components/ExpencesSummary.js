import React from "react";
import { connect } from 'react-redux';
import numeral from 'numeral';
import getVisibleExpenses from "../selectors/expences";
import selectExpencesTotal from  '../selectors/expences-total';

const ExpencesSummary = ({ expenceCount, expencesTotal }) => {
    const exoenceWord = expenceCount === 1 ? "expence" : "expences";
    const formattedExpenceTotal = numeral(expencesTotal/100).format('$0,0.00');
    return(
        <div>
        <h1>Viewing {expenceCount} {exoenceWord}  totalling {formattedExpenceTotal}</h1>
        </div>
    )
}

const mapStateToProps = (state) =>{
    const visibleExpences = getVisibleExpenses(state.expences, state.filters);
    return {
        expenceCount: visibleExpences.length,
        expencesTotal:  selectExpencesTotal(visibleExpences)
    }
};

export default connect(mapStateToProps)(ExpencesSummary)