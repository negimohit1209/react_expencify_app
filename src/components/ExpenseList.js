import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenceListItem';
import selectExpences from '../selectors/expences';
const ExpenseList = (props) => (
    <div>
        <h1>expenceList</h1>
        {props.expences.map((expence) => {
            return <ExpenseListItem key={expence.id} {...expence}/>
        })}
    </div>
)

const mapStateToProps = (state)=>{
    return {
        expences : selectExpences(state.expences, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
