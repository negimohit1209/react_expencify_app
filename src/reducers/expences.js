const expensesReducerdefaultState = [];
const expensesReducer = (state = expensesReducerdefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENCE":
      return [...state, action.expence];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update
          };
        } else {
          return expense;
        }
      });
      case "SET_EXPENCES":
        return action.expences
    default:
      return state;
  }
};

export default expensesReducer;
