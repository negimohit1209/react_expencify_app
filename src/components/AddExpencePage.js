import React from "react";
import ExpenceForm from "./ExpenceForm";
import { connect } from "react-redux";
import { startAddExpence } from "../actions/expences";

class AddExpencePage extends React.Component {
  onSubmit = expence => {
    this.props.startAddExpence(expence);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenceForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpence : (expence) => dispatch(startAddExpence(expence))
}) ;
export default connect(undefined, mapDispatchToProps)(AddExpencePage);
