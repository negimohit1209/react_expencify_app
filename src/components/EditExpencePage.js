import React from "react";
import { connect } from 'react-redux';
import ExpenceForm from './ExpenceForm';
import { editExpense, removeExpence } from '../actions/expences';
const EditExpencePage = (props) => {
    console.log(props);
    return(
        <div>
            <ExpenceForm 
                expence= {props.expence}
                onSubmit = {(expence) => {
                    props.dispatch(editExpense(props.expence.id, expence));
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpence({ id: props.expence.id }));
                props.history.push('/');
              }}>Remove</button>
        </div>        
    )
}

const mapStateToProps = (state, props) => {
    return {
        expence: state.expences.find((expence) =>  expence.id === props.match.params.id)
    }
}
export default connect(mapStateToProps)(EditExpencePage);