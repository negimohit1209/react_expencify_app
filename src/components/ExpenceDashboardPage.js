import React from "react";
import ExpenseList from './ExpenseList';
import ExpenceListFilters from './ExpenceListFilters';
import ExpencesSummary from './ExpencesSummary';
const ExpenceDashboardPage = () => (
    <div>
        <ExpencesSummary />
        <ExpenceListFilters />
        <ExpenseList />
    </div>

)
export default ExpenceDashboardPage;