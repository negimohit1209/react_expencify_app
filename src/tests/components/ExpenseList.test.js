import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expences from '../fixtures/expences';

test('should render ExpenceList with expences', () => {
    const wrapper = shallow(<ExpenseList expences= {expences}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenceList with empty message', () => {
    const wrapper = shallow(<ExpenseList expences= {[]}/>);
    expect(wrapper).toMatchSnapshot();
});