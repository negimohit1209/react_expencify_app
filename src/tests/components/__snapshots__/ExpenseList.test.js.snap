// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenceList with empty message 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`should render ExpenceList with expences 1`] = `
<div>
  <ExpenceListItem
    amount={195}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenceListItem
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenceListItem
    amount={4500}
    createdAt={345600000}
    description="credit card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
